### PR TITLE
Add Users and JWT auth

### PR DESCRIPTION
**Pull request by fisher60**

### CODE REVIEW

**Review by GDWR — CHANGES_REQUESTED** [review]
Please update the `README.md` with the new required env vars and where to put them, i.e. `launchSettings.json`. Please also just create empty values in there to make it easier to spot.

additionally, I am unable to get the database to "update" as I get an error.
```
PS R:\AbandonTech\Reciplease> dotnet ef database update
Build started...
Build succeeded.
System.ArgumentNullException: Google OAuth Client Id cannot be null. (Parameter 'googleClientId')
   at Reciplease.RecipleaseContext..ctor(DbContextOptions`1 options) in R:\AbandonTech\Reciplease\Contexts\RecipleaseContext.cs:line 27
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType
)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass2_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetServiceOrCreateInstance(IServiceProvider provider, Type type)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.<>c__DisplayClass21_3.<FindContextTypes>b__11()
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Google OAuth Client Id cannot be null. (Parameter 'googleClientId')
```

Please provide documentation on how to get the database to match what EF requires so I can run this branch.

File: Contexts/RecipleaseContext.cs [hunk 1]
```diff
@@ -1,15 +1,28 @@
...
+    public string? JwtValidIssuer = Environment.GetEnvironmentVariable("ValidIssuer");
+    public string? JwtSecret = Environment.GetEnvironmentVariable("Secret");
+
     public string DbPath { get; }
 
-    public RecipleaseContext() 
+    public RecipleaseContext(DbContextOptions<RecipleaseContext> options) : base(options)
     {
         var CurrentDirectory = Environment.CurrentDirectory;
         var path = System.IO.Path.Join(CurrentDirectory, "TempData");
```
Comment (GDWR) [inline Contexts/RecipleaseContext.cs:28]: We can use dependency injection to inject the config. Example below for one of the env vars.

```csharp
    private readonly string? _googleClientId;
    
    public RecipleaseContext(DbContextOptions<RecipleaseContext> options, IConfiguration configuration) : base(options)
    {
          googleClientId = configuration["GOOGLE_OAUTH_CLIENT_ID"];
          ... // Other constructor stuff.
    }
```
If we can use `private` and `readonly` as they should only be accessed in this class and set in the contructor.

File: Controllers/UserController.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        _logger = logger;
+        _context = context;
+        _userManager = userManager;
+    }
+
+    public class AuthenticateRequest
+    {
+        [Required]
+        public string IdToken { get; set; }
+    }
```
Comment (GDWR) [inline Controllers/UserController.cs:36]: Find somewhere better to situate this class, I'd recommend in `Models` and a rename to `AuthenticateRequestDto`.

File: Controllers/UserController.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    [AllowAnonymous]
+    [HttpPost("authenticate")]
+    public IActionResult Authenticate([FromBody] AuthenticateRequest data)
+    {
+        GoogleJsonWebSignature.ValidationSettings settings = new GoogleJsonWebSignature.ValidationSettings();
+        
+        settings.Audience = new List<string>() {_context.googleClientId };
+
+        GoogleJsonWebSignature.Payload payload = GoogleJsonWebSignature.ValidateAsync(data.IdToken, settings).Result;
+        return Ok( payload );
```
Comment (GDWR) [inline Controllers/UserController.cs:70]:
```suggestion
        var settings = new GoogleJsonWebSignature.ValidationSettings();
        
        settings.Audience = new List<string> { _context.googleClientId };

        var payload = GoogleJsonWebSignature.ValidateAsync(data.IdToken, settings).Result;
        return Ok( payload );
```

File: Controllers/UserController.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    [HttpPost]
+    [Route("login")]
+    public async Task<IActionResult> Login([FromBody] LoginModel model)
+    {
+        var user = await _userManager.FindByNameAsync(model.Username);
+        
+        if (user != null && await _userManager.CheckPasswordAsync(user, model.Password))
```
Comment (GDWR) [inline Controllers/UserController.cs:79]: Can this be a return early? to reduce nesting.

File: Controllers/UserController.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        {
+            var userRoles = await _userManager.GetRolesAsync(user);
+
+            var authClaims = new List<Claim>
+            {
+                new Claim(ClaimTypes.Name, user.UserName),
+                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
+            };
```
Comment (GDWR) [inline Controllers/UserController.cs:87]: I'm not fully sure about how this works, but I would expect an `email` or something like that to be on here?

Or would that still be accessible in other endpoints, regardless if it is added here?

File: Program.cs [hunk 6]
```diff
@@ -11,11 +16,74 @@
...
+
+    if (clientId == null)
+        throw new ArgumentNullException(nameof(clientId), "Google OAuth Client Id cannot be null.");
+    if (clientSecret == null) 
+        throw new ArgumentNullException(nameof(clientSecret), "Google OAuth Client Secret cannot be null.");
+    
+    googleOptions.ClientId = clientId;
+    googleOptions.ClientSecret = clientSecret;
+});
+
```
Comment (GDWR) [inline Program.cs:57]: indentation looks off.
```suggestion
    {
        var clientId = Environment.GetEnvironmentVariable("GOOGLE_OAUTH_CLIENT_ID");
        var clientSecret = Environment.GetEnvironmentVariable("GOOGLE_OAUTH_CLIENT_SECRET");

        if (clientId == null)
            throw new ArgumentNullException(nameof(clientId), "Google OAuth Client Id cannot be null.");
        if (clientSecret == null) 
            throw new ArgumentNullException(nameof(clientSecret), "Google OAuth Client Secret cannot be null.");
    
        googleOptions.ClientId = clientId;
        googleOptions.ClientSecret = clientSecret;
    });

```

File: Properties/launchSettings.json [hunk 7]
```diff
@@ -16,7 +16,12 @@
       "launchUrl": "swagger",
       "applicationUrl": "https://localhost:7161;http://localhost:5161",
       "environmentVariables": {
-        "ASPNETCORE_ENVIRONMENT": "Development"
+        "ASPNETCORE_ENVIRONMENT": "Development",
+        "GOOGLE_OAUTH_CLIENT_ID": "",
+        "GOOGLE_OAUTH_CLIENT_SECRET": "",
+        "ValidAudience": "",
+        "ValidIssuer": "",
```
Comment (GDWR) [inline Properties/launchSettings.json:23]: We can assume empty Env vars are incorrect for these values, can we add this to where we are checking for `null` so it fails quicker.

**Review by GDWR — CHANGES_REQUESTED** [review]
## Changes

See code review threads and headings below.

### Endpoint to test authorized attribute

Can you make an `Authorized` endpoint that will retrieve some information about your account? For now, a GET method called `Profile` that will return your Username & Email.

### Empty file

`Models/User.cs` is an empty file, remove it.

## Future Notes 

### Error responses
A future issue/pr will have to deal with correctly returning errors to the user, as short passwords give a 500 response.
```
warn: Microsoft.AspNetCore.Identity.UserManager[14]
      User password validation failed: PasswordTooShort;PasswordRequiresNonAlphanumeric;PasswordRequiresDigit;PasswordRequiresUpper.
```
### Missing XML Docstrings

We are missing XML Docstrings for the endpoints, this can be deferred to a future PR/Issue so we can complete issues at the same time.

File: Contexts/RecipleaseContext.cs (hunk 1 above)
Comment (fisher60) [reply Contexts/RecipleaseContext.cs:28]: Yes, we should add this to a future refactor in my opinion.

File: Controllers/UserController.cs (hunk 5 above)
Comment (fisher60) [reply Controllers/UserController.cs]: The current setup could allow user data to be retrieved via username at the `profile` endpoint if needed, however we should not have the need to get user data from the JWT. The JWT should pass user roles in the token, no DB queries should need to be made when using or validating a JWT.

File: Controllers/AuthController.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        var token = GetToken(authClaims);
+
+        return Ok(new
+        {
+            token = new JwtSecurityTokenHandler().WriteToken(token),
+            expiration = token.ValidTo
+        });
+
+            return Unauthorized();
```
Comment (GDWR) [inline Controllers/AuthController.cs:71]: un reachable code?

File: Controllers/UserController.cs [hunk 9]
```diff
@@ -54,66 +42,15 @@ public async Task<IActionResult> Register([FromBody] RegisterModel model)
...
+            return StatusCode(500, result.Errors);
+        return CreatedAtAction(nameof(Profile), new {Username = user.UserName});
     }
 
-    [HttpPost]
-    [Route("login")]
-    public async Task<IActionResult> Login([FromBody] LoginModel model)
+    [HttpGet]
+    [Route("profile")]
+    public async Task<ActionResult<UserDto>> Profile([FromQuery] UserInputDto user)
```
Comment (GDWR) [inline Controllers/UserController.cs:51]: This should be an `[Authorized]` endpoint, where the correct profile is gathered from the JWT as that is how we authorize them.

**Review by GDWR — CHANGES_REQUESTED** [review]
The profile endpoint you have made doesn't test/use the auth you have made. Please make it do so.

File: Contexts/RecipleaseContext.cs (hunk 1 above)
Comment (GDWR) [reply Contexts/RecipleaseContext.cs:28]: Please make a chasing issue for this, after merge.

File: Controllers/UserController.cs (hunk 9 above)
Comment (fisher60) [reply Controllers/UserController.cs:51]: All user endpoints are currently authorized, the register endpoint is set to explicitly allow unauthorized. So the profile endpoint is authorized by default.

File: Controllers/AuthController.cs (hunk 8 above)
Comment (fisher60) [reply Controllers/AuthController.cs]: Oops, remnant of the old method.

**Review by GDWR — APPROVED** [review]

**Comment by GDWR** [comment]
@fisher60 Please create the new Issues mentioned across this ticket. Link them here too for accessibility.

**Comment by fisher60** [comment]
issues #10, #11, and #12 have been created as a result of the work here.

**Pull request merged**